src/util/geom.ts

```diff
@@ -478,3 +482,45 @@ export function getViewFrustrumWorldBounds(
     }
   }
 }
+
+export function calculateOrientedSliceScales(
+  orientation: quat | undefined,
+  scales: vec3,
+  units: readonly string[],
+  tolerance: number = 1e-6,
+): OrientedSliceScales | null {
+  function extractContributingScales(
+    matrixRow: 0 | 1 | 2,
+  ): { scale: number; unit: string } | null {
+    const indices = [0, 1, 2].map((i) => i + 3 * matrixRow);
+    const contributingScales: number[] = [];
+    const contributingUnits: string[] = [];
+
+    for (const index of indices) {
```

Comment:
```
for (let i = 0; i < 3; ++i) {
  let index = i + 3 * matrixRow;
  ...
}
```

Rather than use an array / set, you can just use:

```
let contributingScale: number | undefined;
let contributingUnit: number | undefined;
```

and then check if the current unit / scale matche the previous one if not undefined.

For units the exact comparison you have makes sense.  For the scales it would probably be better to use an approximate comparison, i.e. `Math.abs(a - b) / Math.max(a, b) < 1e-6` in case the scales are not exactly equal due to some numerical issue (not sure how likely that is, but may as well be defensive).